src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs

```diff
@@ -11,73 +11,150 @@ namespace Microsoft.CodeAnalysis.Differencing
     /// </summary>
     internal abstract class LongestCommonSubsequence<TSequence>
     {
-        private const int DeleteCost = 1;
-        private const int InsertCost = 1;
-        private const int UpdateCost = 2;
-
-        protected abstract bool ItemsEqual(TSequence oldSequence, int oldIndex, TSequence newSequence, int newIndex);
-
-        protected IEnumerable<KeyValuePair<int, int>> GetMatchingPairs(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        // VArray class enables array indexing in range [-d...d].
+        private class VArray
         {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
+            private int[] array;
+            private int offset;
 
-            while (i != 0 && j != 0)
+            public VArray(int d, VArray previousVArray) : this(d)
```

Comment:
Since ```d``` is not used in the body it would be better to turn this constructor into an instance method (e.g. ```CopyFrom(VArray previous)```) and change the call-sites from ```new VArray(d, previous) to ```new VArray(d).CopyFrom(previous)```.